runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,431 @@
+     * @return keyset of the table
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        applyWritesForReadOnTable(table);
+        table.getMetrics().numKeySets.incrementAndGet();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param table Table< K, V, M > object on which the scan must be done.
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        applyWritesForReadOnTable(table);
+        table.getMetrics().numScans.incrementAndGet();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Apply all pending writes (if any) to serve read queries.
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void applyWritesForReadOnTable(Table<K, V, M> tableBeingRead) {
+        tablesInTxn.add(tableBeingRead);
+        txnType |= READ_ONLY;
+        if (!operations.isEmpty()) {
+            operations.forEach(Runnable::run);
+            operations.clear();
+        }
+    }
+
+    /**
+     * Start the actual corfu transaction. Ensure there isn't one already in the same thread.
+     */
+    private void txBeginInternal() {
+        if (TransactionalContext.isInTransaction()) {
+            log.error("Cannot start new transaction in this thread without ending previous one");
+            throw new TransactionAlreadyStartedException(TransactionalContext.getRootContext().toString());
+        }
+        this.txnStartTime = System.nanoTime();
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Commit the transaction.
+     * For a transaction that only has write operations, this method will start and end
+     * a corfu transaction keeping the "critical" section small while batching up all updates.
+     * For a transaction that has reads, this will end the transaction.
+     * The commit returns successfully if the write transaction was committed.
+     * Otherwise this throws a TransactionAbortedException with a cause.
+     * The cause and the caller's intent of the transaction can determine if this aborted
+     * Transaction can be retried.
+     * @return - address at which the commit of this transaction occurred.
+     */
+    public long commit() {
```

Comment:
Can we add a sanity check in commit() that checks if we are still inside the transcation context to prevent the case that commit is called twice or it's called outside of try finally block?

Reply:
done.